### PR TITLE
fix(Collision): stop collisions when either side is disabled - fixes ExtendRealityLtd/VRTK#2005

### PR DESCRIPTION
**Pull request by thestonefox**
There is an issue in Unity where when two objects are colliding will
create a collision started event, but whilst the objects are
intersecting, if one of the objects is disabled then Unity does not
consider this to stop the collision between the two objects so no
OnCollisionExit or OnTriggerExit occurs. However, if the disabled
object is re-enabled and is still in the same intersecting position
then Unity will consider this a new Enter event.

This causes a problem for the CollisionTracker as it means the state
of the tracker can become out of sync. If a collision starts twice
but no stop was called between, then the tracked collision state is
not valid. This would then have knock on effects in usages such as
the ActiveCollisionsContainer where a started collision is treated
as an active collision, but then if the object is disabled then
no stopped collision occurs and therefore that would still be
considered as an active collision.

To fix this issue and work around the shortcomings of Unity, a new
component is added at runtime to the GameObject that contains the
collider that is colliding with the CollisionTracker. This new
component then monitors the disabled state of the GameObject and upon
that GameObject becoming disabled will stop the collision manually
on the linked CollisionTracker. It will also then remove any traces
of the magically created component.

The CollisionTracker now also tracks any colliders it may be
intersecting with and therefore if the CollisionTracker is disabled
then any collider it may be linked to also has the Collision Stopped
event emitted for to ensure both sides of the collision adhere to the
disabled state.

This new functionality is optional and can be turned off by the `Stop
Collisions On Disable` property, which can only be set at edit time to
prevent fundamental state getting out of sync if it was changed at
runtime.

The `AddComponent` and `Destroy` are used over a pooling method because
these only occur on the `Enter/Exit` events so are not firing all the
time and therefore performance hits should be reduced. But also,
pooling would create a large set of GameObjects for each
CollisionTracker and they would all have to independently manage these
pools themselves resulting in a potential large number of pooled
GameObjects being created on scene start, which would have a potential
much larger impact on initial performance.

The tests have been updated to showcase the issue of disabling a
GameObject on either side of the collision now can successfully result
in the Stopped event being emitted.

### CODE REVIEW

File: Runtime/Tracking/Collision/CollisionTracker.cs [hunk 1]
```diff
@@ -1,19 +1,66 @@
...
+        [field: DocumentedByXml, Restricted(RestrictedAttribute.Restrictions.ReadOnlyAtRunTime)]
+        public bool StopCollisionsOnDisable { get; protected set; } = true;
+
+        /// <summary>
+        /// A collection of current existing collisions.
+        /// </summary>
+        protected List<Collider> trackedCollisions = new List<Collider>();
+
+        /// <summary>
+        /// Stops the collision between this <see cref="CollisionTracker"/> and the given <see cref="Collider"/>. If there is still a physical intersection, then the collision will start again next physics frame.
```
Comment (bddckr) [inline Runtime/Tracking/Collision/CollisionTracker.cs]:
```suggestion
        /// Stops the collision between this <see cref="CollisionTracker"/> and the given <see cref="Collider"/>. If there is still a physical intersection, then the collision will start again in the next physics frame.
```

File: Runtime/Tracking/Collision/CollisionTracker.cs [hunk 2]
```diff
@@ -1,19 +1,66 @@
...
+    using Malimbe.PropertySerializationAttribute;
+    using Zinnia.Data.Attribute;
 
     /// <summary>
     /// Tracks collisions on the <see cref="GameObject"/> this component is on.
     /// </summary>
     public class CollisionTracker : CollisionNotifier
     {
+        /// <summary>
+        /// Causes collisions to stop if the <see cref="GameObject"/> either side of the collision is disabled.
```
Comment (bddckr) [inline Runtime/Tracking/Collision/CollisionTracker.cs]:
```suggestion
        /// Causes collisions to stop if the <see cref="GameObject"/> on either side of the collision is disabled.
```

File: Runtime/Tracking/Collision/CollisionTracker.cs [hunk 3]
```diff
@@ -1,19 +1,66 @@
...
+        }
+
         protected virtual void OnCollisionEnter(Collision collision)
         {
             if ((StatesToProcess & CollisionStates.Enter) == 0)
             {
                 return;
             }
 
+            AddTrackedNode(collision.collider);
```
Comment (bddckr) [inline Runtime/Tracking/Collision/CollisionTracker.cs]: Move these calls (`AddTrackedNode` and `RemoveTrackedNode`) into the `On*` (e.g. this one's `OnCollisionStarted`) methods we call here. I.e. override the methods in this subclass.

Since the `On*` methods are our abstraction it feels like it makes more sense to do this additional logic in them rather than here, just for the sake of ensuring it's done when one would expect it to. This also ensure subclasses don't break that functionality when calling the `On*` methods themselves.

File: Runtime/Tracking/Collision/CollisionTracker.cs [hunk 4]
```diff
@@ -64,7 +113,87 @@ protected virtual void OnTriggerExit(Collider collider)
...
 
+            RemoveTrackedNode(collider);
             OnCollisionStopped(eventData.Set(this, true, null, collider));
         }
+
+        /// <summary>
+        /// Adds a <see cref="CollisionTrackerDisabledObserver"/> to the <see cref="GameObject"/> that contains the <see cref="Collider"/> causing the collision.
+        /// </summary>
+        /// <param name="target">The target to add the <see cref="CollisionTrackerDisabledObserver"/> component to.</param>
+        protected virtual void AddTrackedNode(Collider target)
```
Comment (bddckr) [inline Runtime/Tracking/Collision/CollisionTracker.cs]:
```suggestion
        protected virtual void AddDisabledObserver(Collider target)
```

File: Runtime/Tracking/Collision/CollisionTracker.cs [hunk 5]
```diff
@@ -64,7 +113,87 @@ protected virtual void OnTriggerExit(Collider collider)
...
+        {
+            if (target == null || !StopCollisionsOnDisable)
+            {
+                return;
+            }
+
+            trackedCollisions.Add(target);
+            CollisionTrackerDisabledObserver node = target.gameObject.AddComponent<CollisionTrackerDisabledObserver>();
```
Comment (bddckr) [inline Runtime/Tracking/Collision/CollisionTracker.cs]:
```suggestion
            CollisionTrackerDisabledObserver observer = target.gameObject.AddComponent<CollisionTrackerDisabledObserver>();
```

File: Runtime/Tracking/Collision/CollisionTracker.cs [hunk 6]
```diff
@@ -64,7 +113,87 @@ protected virtual void OnTriggerExit(Collider collider)
...
+                    trackedCollisions.Remove(target);
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Observes the disabled state of any <see cref="GameObject"/> that the <see cref="CollisionTracker"/> is currently colliding with.
+    /// </summary>
+    public class CollisionTrackerDisabledObserver : MonoBehaviour
```
Comment (bddckr) [inline Runtime/Tracking/Collision/CollisionTracker.cs]: Put in standalone file instead?

File: Runtime/Tracking/Collision/CollisionTracker.cs [hunk 7]
```diff
@@ -64,7 +113,87 @@ protected virtual void OnTriggerExit(Collider collider)
...
+            CollisionTrackerDisabledObserver node = target.gameObject.AddComponent<CollisionTrackerDisabledObserver>();
+            node.Source = this;
+            node.Target = target;
+        }
+
+        /// <summary>
+        /// Removes the <see cref="CollisionTrackerDisabledObserver"/> from the <see cref="GameObject"/> that contains the <see cref="Collider"/> ceasing the collision.
+        /// </summary>
+        /// <param name="target">The target to remove the <see cref="CollisionTrackerDisabledObserver"/> component from.</param>
+        protected virtual void RemoveTrackedNode(Collider target)
```
Comment (bddckr) [inline Runtime/Tracking/Collision/CollisionTracker.cs]:
```suggestion
        protected virtual void RemoveDisabledObserver(Collider target)
```

File: Runtime/Tracking/Collision/CollisionTracker.cs [hunk 8]
```diff
@@ -64,7 +113,87 @@ protected virtual void OnTriggerExit(Collider collider)
...
+        {
+            if (target == null || !StopCollisionsOnDisable)
+            {
+                return;
+            }
+
+            foreach (CollisionTrackerDisabledObserver node in target.gameObject.GetComponents<CollisionTrackerDisabledObserver>().ToArray())
```
Comment (bddckr) [inline Runtime/Tracking/Collision/CollisionTracker.cs]:
```suggestion
            foreach (CollisionTrackerDisabledObserver observer in target.gameObject.GetComponents<CollisionTrackerDisabledObserver>())
```
Not sure why that `.ToArray` was used, but it seems unnecessary.

File: Runtime/Tracking/Collision/CollisionTracker.cs [hunk 9]
```diff
@@ -64,7 +113,87 @@ protected virtual void OnTriggerExit(Collider collider)
...
+
+        /// <summary>
+        /// Whether <see cref="this"/> is being destroyed.
+        /// </summary>
+        protected bool isDestroyed;
+
+        /// <summary>
+        /// Destroys <see cref="this"/> from the scene.
+        /// </summary>
+        public virtual void Destroy()
```
Comment (bddckr) [inline Runtime/Tracking/Collision/CollisionTracker.cs]: Shouldn't this instead be `OnDestroyed`? What happens if a user destroys this component, even if it's because of destruction of a parent game object? This logic should still run, shouldn't it?

I may be missing some details around how this magic method works, I know you brought it up on Slack...

File: Runtime/Tracking/Collision/CollisionTracker.cs [hunk 10]
```diff
@@ -64,7 +113,87 @@ protected virtual void OnTriggerExit(Collider collider)
...
+        {
+            isDestroyed = true;
+            Destroy(this);
+        }
+
+        protected virtual void OnDisable()
+        {
+            if (!isDestroyed)
```
Comment (bddckr) [inline Runtime/Tracking/Collision/CollisionTracker.cs]: Why is this needed to be checked?

File: Runtime/Tracking/Collision/CollisionTracker.cs (hunk 3 above)
Comment (thestonefox) [reply Runtime/Tracking/Collision/CollisionTracker.cs]: Well this fix is actually fixing the issue Unity has with `OnCollisionXXX` and `OnTriggerXXX` what if there was a different implementation that extended the `CollisionNotifier`? You may not even need this work around.

I kinda feel like the `CollisionNotifier` is super generic and the `CollisionTracker` is our implementation of how Unity tracks collisions (badly).

Which would be my case for keeping these only in here and not dirtying the `CollisionNotifier`..... thoughts?

File: Runtime/Tracking/Collision/CollisionTracker.cs (hunk 8 above)
Comment (thestonefox) [reply Runtime/Tracking/Collision/CollisionTracker.cs]: `ToArray` was just a copy/paste from another foreach. I've deleted it as it isn't needed.

File: Runtime/Tracking/Collision/CollisionTracker.cs (hunk 6 above)
Comment (thestonefox) [reply Runtime/Tracking/Collision/CollisionTracker.cs]: Well if it's in a standalone file it can be dragged into the inspector, which we don't really want. It's basically the "magic" helper for this CollisionTracker.

File: Runtime/Tracking/Collision/CollisionTracker.cs (hunk 9 above)
Comment (thestonefox) [reply Runtime/Tracking/Collision/CollisionTracker.cs]: This isn't the Unity magic method which is `OnDestroy`

This is a custom method that destroys the observer component.

It's so the `CollisionTracker` can remove the component when the collision ends.

The reason it has the `isDestroyed` flag is so I know the component has been removed manually and therefore I don't want the `Source.StopCollision()` method firing when the component is disabled.

Because destroying the object will call `OnDisable` and then you'd get two stop messages..

If you destroy the object that this observer component is on, then `Destroy` isn't called and doesn't set `isDestroyed` so you *do* get the `OnDisable` called and therefore it does correctly call the `StopCollision`

File: Runtime/Tracking/Collision/CollisionTracker.cs (hunk 10 above)
Comment (thestonefox) [reply Runtime/Tracking/Collision/CollisionTracker.cs]: See the explanation above

File: Runtime/Tracking/Collision/CollisionTracker.cs (hunk 3 above)
Comment (bddckr) [reply Runtime/Tracking/Collision/CollisionTracker.cs]: Oh that's true. Anyone else using the `On*` methods should have proper ways to work with them right away, and if they don't they most likely are using this subclass anyway, so it wouldn't matter how we do it. As long as it's happening starting from Unity's magic callback methods that is, which this solution does.

Closing as resolved :)

File: Runtime/Tracking/Collision/CollisionTracker.cs (hunk 6 above)
Comment (bddckr) [reply Runtime/Tracking/Collision/CollisionTracker.cs]: `[AddComponentMenu("")]` should hide it from there! I think we even use that elsewhere already :)

Comment (thestonefox) [reply Runtime/Tracking/Collision/CollisionTracker.cs]: It hides it from the drop down but doesn't prevent you from dragging that component into the inspector window, which we still dont want.

File: Runtime/Tracking/Collision/CollisionTracker.cs (hunk 9 above)
Comment (bddckr) [reply Runtime/Tracking/Collision/CollisionTracker.cs]: I looked into this some more and there is indeed no way to say "OnDisable, but only when *not* destroyed". Too bad Unity always throws in some weird specific implementation detail.

File: Runtime/Tracking/Collision/CollisionTracker.cs (hunk 6 above)
Comment (bddckr) [reply Runtime/Tracking/Collision/CollisionTracker.cs]: 🤷‍♂ There's only so much I'd do to prevent obvious mistakes.

I just don't like relying on this Unity behavior as it seems undocumented, or at least not documented as a "feature" that is intended this way..
